src/psyclone/core/component_indices.py

```diff
@@ -0,0 +1,142 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2021, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+'''This module provides a class to manage indices in variable accesses.'''
+
+from __future__ import print_function, absolute_import
+
+
+from psyclone.errors import InternalError
+
+
+class ComponentIndices(object):
+    '''This class stores index information for variable accesses. It stores
+    one index list for each component of a variable, e.g. for `a(i)%b(j)`
+    it would store `[ [i], [j] ]`. Even for scalar accesses an emtpy list
+    is stored, so `a` would have the component indices `[ [] ]`, and `a%b`
+    would have `[ [], [] ]`.
+
+    As a shortcut, the `indices` parameter can be None or an empty list
+    (which then creates the component indices as `[[]]`), a list `l` (which
```

Comment:
I've reworded that - check if this is better. The issue is that the component indices does not know anything about the corresponding signature. In the access_info code that is handled (and indeed there it becomes the last component), here it is only a shortcut for non-structured variables, so I want to avoid 'last component' (while this is correct, it would be misleading).

Reply:
Yes, I think that's better - thanks.